src/impl/tlstransport.cpp

```diff
@@ -292,7 +292,223 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+
+void TlsTransport::postHandshake() {
+	// Dummy
+}
+
+void TlsTransport::doRecv() {
+	std::lock_guard lock(mRecvMutex);
+	--mPendingRecvCount;
+
+	if (state() != State::Connecting && state() != State::Connected)
+		return;
+
+	try {
+		const size_t bufferSize = 4096;
+		char buffer[bufferSize];
+
+		// Handle handshake if connecting
+		if (state() == State::Connecting) {
+			auto handshakeStatus = mbedtls_ssl_handshake(&mSsl);
+			switch (handshakeStatus) {
+				case 0:
+					break;
+				case MBEDTLS_ERR_SSL_WANT_WRITE:
+				case MBEDTLS_ERR_SSL_WANT_READ:
+				case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+				case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS: {
+					auto timeout = milliseconds(100);
+					ThreadPool::Instance().schedule(timeout, [weak_this = weak_from_this()]() {
+						if (auto locked = weak_this.lock())
+							locked->doRecv();
+					});
+					return;
+				}
+				default:
+				mbedtls::check(handshakeStatus);
+			}
+
+			PLOG_INFO << "TLS handshake finished";
+			changeState(State::Connected);
+			postHandshake();
+		}
+
+		if (state() == State::Connected) {
+			while (true) {
+				auto ret = mbedtls_ssl_read(&mSsl, (unsigned char *) buffer, bufferSize);
+				if (ret == 0 || ret == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+					// Closed
+					PLOG_DEBUG << "TLS connection cleanly closed";
+					break;
+				}
+				switch (ret) {
+					case MBEDTLS_ERR_SSL_WANT_READ:
+					case MBEDTLS_ERR_SSL_WANT_WRITE:
+					case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+					case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS:
+					case MBEDTLS_ERR_SSL_CLIENT_RECONNECT:
+						return;
+				}
+				mbedtls::check(ret);
+
+				auto *b = reinterpret_cast<byte *>(buffer);
+				recv(make_message(b, b + ret));
+			}
+		}
+	} catch (const std::exception &e) {
+		PLOG_ERROR << "TLS recv: " << e.what();
+	}
+
+	PLOG_INFO << "TLS closed";
+	changeState(State::Disconnected);
+	recv(nullptr);
+}
+
+int TlsTransport::WriteCallback(void *ctx, const unsigned char *buf, size_t len) {
+	auto *t = static_cast<TlsTransport *>(ctx);
+	auto *b = reinterpret_cast<const byte *>(buf);
+	if (!t->outgoing(make_message(b, b + len))) {
+		return MBEDTLS_ERR_SSL_WANT_READ;
+	}
+
+	return len;
+}
+
+int TlsTransport::ReadCallback(void *ctx, unsigned char *buf, size_t len) {
+	TlsTransport *t = static_cast<TlsTransport *>(ctx);
+	try {
+		message_ptr &message = t->mIncomingMessage;
+		size_t &position = t->mIncomingMessagePosition;
+
+		if (message && position >= message->size())
+			message.reset();
+
+		if (!message) {
+			position = 0;
+			while (auto next = t->mIncomingQueue.pop()) {
+				message = *next;
+				if (message->size() > 0)
+					break;
+				else
+					t->recv(message); // Pass zero-sized messages through
+			}
+		}
+
+		if (message) {
+			size_t available = message->size() - position;
+			ssize_t writeLen = std::min((unsigned long) len, available);
+			std::memcpy(buf, message->data() + position, writeLen);
+			position += len;
+			return len;
```

Comment:
I think the wrong size if added and returned.
```suggestion
			size_t available = message->size() - position;
			size_t writeLen = std::min(len, available);
			std::memcpy(buf, message->data() + position, writeLen);
			position += writeLen;
			return writeLen;
```